File: ailake-go/rabitq.go
// SPDX-License-Identifier: MIT OR Apache-2.0
// Copyright (c) 2026 Thiago Egon Lange
// RaBitQ flat index: deserializer and brute-force search.
//
// Wire format: bincode v1 serialization of ailake_index::RaBitQIndex:
//   codebook.dim:    usize  (u64 LE)
//   codebook.seed:   u64    (projection matrix is regenerated, not stored)
//   entries:         Vec<RaBitQVec>
//     each entry:
//       code:  Vec<u8>   (packed sign bits, ceil(dim/8) bytes)
//       norm:  f32
//       scale: f32
//   row_ids: Vec<u64>
//   metric:  u32          (enum variant: 0=cosine, 1=euclidean, 2=dot, 3=normalized-cosine)
//   dim:     u32
//   raw_f16: Option<Vec<u16>>  (f16 bits; None=0x00, Some=0x01+slice)
package ailake

import (
	"fmt"
	"math"
	"math/bits"
	"math/rand"
	"sort"
)

// RaBitQVec is the binary-quantized representation of one database vector.
type RaBitQVec struct {
	Code  []byte  // packed sign bits: bit i = sign(P·x̂)[i]
	Norm  float32 // original L2 norm
	Scale float32 // sum(|P·x̂|) / sqrt(dim)
}

// RaBitQIndex is a deserialized AI-Lake RaBitQ flat index ready for search.
type RaBitQIndex struct {
	Seed    uint64
	Dim     uint32
	Metric  uint8
	Entries []RaBitQVec
	RowIDs  []uint64
	RawF16  []uint16 // nil when keep_raw=false; length = len(Entries)*Dim
	proj    []float32 // dim×dim, row-major; regenerated from Seed
}

// DeserializeRaBitQ deserializes a bincode-encoded RaBitQIndex from buf.
// The projection matrix is regenerated from the stored seed (not stored on disk).
func DeserializeRaBitQ(buf []byte) (*RaBitQIndex, error) {
	r := newBincodeReader(buf)

	dim, err := r.readUsize()
	if err != nil {
		return nil, fmt.Errorf("rabitq: read codebook.dim: %w", err)
	}
	seed, err := r.readU64()
	if err != nil {
		return nil, fmt.Errorf("rabitq: read codebook.seed: %w", err)
	}

	entryCount, err := r.readUsize()
	if err != nil {
		return nil, fmt.Errorf("rabitq: read entries length: %w", err)
	}
	entries := make([]RaBitQVec, entryCount)
	for i := range entries {
		codeLen, err := r.readUsize()
		if err != nil {
			return nil, fmt.Errorf("rabitq: entry %d code length: %w", i, err)
		}
		code, err := r.readN(int(codeLen))
		if err != nil {
			return nil, fmt.Errorf("rabitq: entry %d code: %w", i, err)
		}
		code2 := make([]byte, len(code))
		copy(code2, code)
		norm, err := r.readF32()
		if err != nil {
			return nil, fmt.Errorf("rabitq: entry %d norm: %w", i, err)
		}
		scale, err := r.readF32()
		if err != nil {
			return nil, fmt.Errorf("rabitq: entry %d scale: %w", i, err)
		}
		entries[i] = RaBitQVec{Code: code2, Norm: norm, Scale: scale}
	}

	rowIDs, err := r.readU64Slice()
	if err != nil {
		return nil, fmt.Errorf("rabitq: read row_ids: %w", err)
	}

	metricVariant, err := r.readU32()
	if err != nil {
		return nil, fmt.Errorf("rabitq: read metric: %w", err)
	}

	dim32, err := r.readU32()
	if err != nil {
		return nil, fmt.Errorf("rabitq: read dim: %w", err)
	}

	// Option<Vec<f16>>: tag 0x00=None, 0x01=Some
	optTag, err := r.readU8()
	if err != nil {
		return nil, fmt.Errorf("rabitq: read raw_f16 tag: %w", err)
	}
	var rawF16 []uint16
	if optTag == 1 {
		n, err := r.readUsize()
		if err != nil {
			return nil, fmt.Errorf("rabitq: read raw_f16 length: %w", err)
		}
		rawF16 = make([]uint16, n)
		for i := range rawF16 {
			v, err := r.readN(2)
			if err != nil {
				return nil, fmt.Errorf("rabitq: raw_f16[%d]: %w", i, err)
			}
			rawF16[i] = uint16(v[0]) | uint16(v[1])<<8
		}
	}

	idx := &RaBitQIndex{
		Seed:    seed,
		Dim:     dim32,
		Metric:  uint8(metricVariant),
		Entries: entries,
		RowIDs:  rowIDs,
		RawF16:  rawF16,
	}
	idx.buildProj(int(dim))
	return idx, nil
}

// buildProj regenerates the dim×dim projection matrix from the seed.
// Uses the same algorithm as RaBitQCodebook::rebuild_proj in Rust:
// column-normalized random Gaussian matrix.
func (idx *RaBitQIndex) buildProj(dim int) {
	proj := make([]float32, dim*dim)
	rng := rand.New(rand.NewSource(int64(idx.Seed)))
	for col := 0; col < dim; col++ {
		var normSq float64
		for row := 0; row < dim; row++ {
			v := rng.Float64()*2 - 1
			proj[row*dim+col] = float32(v)
			normSq += v * v
		}
		inv := float32(1.0 / math.Sqrt(normSq+1e-24))
		for row := 0; row < dim; row++ {
			proj[row*dim+col] *= inv
		}
	}
	idx.proj = proj
}

// project applies the rotation matrix P to v, returning P·v.
func (idx *RaBitQIndex) project(v []float32) []float32 {
	dim := int(idx.Dim)
	out := make([]float32, dim)
	for i := 0; i < dim; i++ {
		var s float32
		row := idx.proj[i*dim : (i+1)*dim]
		for j, x := range row {
			s += x * v[j]
		}
		out[i] = s
	}
	return out
}

// bitsFromSigns packs sign bits: bit i = (v[i] > 0).
func bitsFromSigns(v []float32) []byte {
	codeLen := (len(v) + 7) / 8
	code := make([]byte, codeLen)
	for i, val := range v {
		if val > 0 {
			code[i/8] |= 1 << uint(i&7)
		}
	}
	return code
}

// hammingXorPopcount counts XOR bits between two equal-length byte slices.
func hammingXorPopcount(a, b []byte) int {
	var total int
	for i := range a {
		total += bits.OnesCount8(a[i] ^ b[i])
	}
	return total
}

func f16ToF32Go(bits uint16) float32 {
	return f16ToF32(bits)
}

// Search returns top-k (rowID, distance) pairs sorted by distance ascending.
// rerank: if true and RawF16 is set, reranks top 3×k candidates exactly.
func (idx *RaBitQIndex) Search(query []float32, topK int, rerank bool) []SearchResult {
	if len(idx.Entries) == 0 {
		return nil
	}
	dim := int(idx.Dim)

	// Normalize query
	var qNorm float64
	for _, x := range query {
		qNorm += float64(x) * float64(x)
	}
	qNorm = math.Sqrt(qNorm)
	qHat := make([]float32, dim)
	if qNorm > 1e-12 {
		for i, x := range query {
			qHat[i] = float32(float64(x) / qNorm)
		}
	} else {
		copy(qHat, query)
	}

	// Project query
	qProj := idx.project(qHat)
	var qScaleSum float64
	for _, x := range qProj {
		if x < 0 {
			qScaleSum -= float64(x)
		} else {
			qScaleSum += float64(x)
		}
	}
	qScale := float32(qScaleSum / math.Sqrt(float64(dim)))
	qCode := bitsFromSigns(qProj)

	type scored struct {
		idx  int
		dist float32
	}

	all := make([]scored, len(idx.Entries))
	for i, e := range idx.Entries {
		hamming := hammingXorPopcount(qCode, e.Code)
		ip := (1.0 - 2.0*float32(hamming)/float32(dim)) * qScale * e.Scale

		var dist float32
		switch idx.Metric {
		case MetricCosine, MetricNormalizedCosine:
			dist = 1.0 - ip
		case MetricDotProduct:
			dist = -ip * float32(qNorm) * e.Norm
		default: // Euclidean
			normX := e.Norm
			d2 := float32(qNorm*qNorm) + normX*normX - 2*ip*float32(qNorm)*normX
			if d2 < 0 {
				d2 = 0
			}
			dist = float32(math.Sqrt(float64(d2)))
		}
		all[i] = scored{i, dist}
	}

	sort.Slice(all, func(a, b int) bool { return all[a].dist < all[b].dist })

	// Reranking with raw F16 vectors
	candidates := topK
	if rerank && len(idx.RawF16) > 0 {
		candidates = topK * 3
	}
	if candidates > len(all) {
		candidates = len(all)
	}
	top := all[:candidates]

	if rerank && len(idx.RawF16) > 0 {
		reranked := make([]scored, len(top))
		for j, s := range top {
			dbF16 := idx.RawF16[s.idx*dim : (s.idx+1)*dim]
			dbF32 := make([]float32, dim)
			for k, bits := range dbF16 {
				dbF32[k] = f16ToF32Go(bits)
			}
			var dist float32
			switch idx.Metric {
			case MetricCosine, MetricNormalizedCosine:
				dist = cosineDistGo(query, dbF32)
			case MetricDotProduct:
				dist = dotDistGo(query, dbF32)
			default:
				dist = euclideanDistGo(query, dbF32)
			}
			reranked[j] = scored{s.idx, dist}
		}
		sort.Slice(reranked, func(a, b int) bool { return reranked[a].dist < reranked[b].dist })
		top = reranked
	}

	if topK > len(top) {
		topK = len(top)
	}
	out := make([]SearchResult, topK)
	for i, s := range top[:topK] {
		out[i] = SearchResult{RowID: idx.RowIDs[s.idx], Distance: s.dist}
	}
	return out
}

func cosineDistGo(a, b []float32) float32 {
	var dot, na, nb float64
	for i := range a {
		dot += float64(a[i]) * float64(b[i])
		na += float64(a[i]) * float64(a[i])
		nb += float64(b[i]) * float64(b[i])
	}
	denom := math.Sqrt(na) * math.Sqrt(nb)
	if denom < 1e-12 {
		return 1
	}
	return float32(1 - dot/denom)
}

func dotDistGo(a, b []float32) float32 {
	var dot float64
	for i := range a {
		dot += float64(a[i]) * float64(b[i])
	}
	return float32(-dot)
}

func euclideanDistGo(a, b []float32) float32 {
	var sum float64
	for i := range a {
		d := float64(a[i]) - float64(b[i])
		sum += d * d
	}
	return float32(math.Sqrt(sum))
}

File: ailake-index/src/rabitq.rs
// SPDX-License-Identifier: MIT OR Apache-2.0
//! RaBitQ flat index: brute-force search over binary codes with optional reranking.
//!
//! Search algorithm:
//!   1. Apply random rotation to query (once per search).
//!   2. For each database vector: XOR + popcount → IP estimate (O(dim/8)).
//!   3. Return top-k by estimated distance.
//!   4. Optional reranking: exact F16 distance for top `rerank_factor × k` candidates.
//!
//! Use case: very compact index (1 bit/dim + 8 bytes/vector overhead) with
//! better recall than naive binary quantization at the same storage cost.

use ailake_core::{AilakeError, AilakeResult, RowId, VectorMetric};
use ailake_vec::{
    exact_distance,
    rabitq::{bits_from_signs, encode_batch, RaBitQCodebook, RaBitQVec},
};
use half::f16;
use serde::{Deserialize, Serialize};

// ── Index ─────────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RaBitQConfig {
    /// Seed for the random rotation matrix. Regenerated at load time — not stored.
    pub seed: u64,
    /// Keep raw F16 vectors for exact reranking. Default true.
    #[serde(default = "default_keep_raw")]
    pub keep_raw: bool,
}

fn default_keep_raw() -> bool {
    true
}

impl Default for RaBitQConfig {
    fn default() -> Self {
        Self {
            seed: 0,
            keep_raw: true,
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RaBitQIndex {
    pub codebook: RaBitQCodebook,
    pub entries: Vec<RaBitQVec>,
    pub row_ids: Vec<u64>,
    pub metric: VectorMetric,
    pub dim: u32,
    /// Raw F16 vectors for reranking (present when `keep_raw = true`).
    pub raw_f16: Option<Vec<f16>>,
}

impl RaBitQIndex {
    /// Build index from raw F32 vectors.
    pub fn build(
        row_ids: &[RowId],
        vectors: &[Vec<f32>],
        metric: VectorMetric,
        config: RaBitQConfig,
        keep_raw: bool,
    ) -> AilakeResult<Self> {
        if vectors.is_empty() {
            return Err(AilakeError::InvalidArgument(
                "RaBitQIndex::build requires at least one vector".into(),
            ));
        }
        let dim = vectors[0].len();
        let cb = RaBitQCodebook::new(dim, config.seed);
        let entries = encode_batch(&cb, vectors);

        let raw_f16 = if keep_raw {
            Some(
                vectors
                    .iter()
                    .flat_map(|v| v.iter().map(|&x| f16::from_f32(x)))
                    .collect(),
            )
        } else {
            None
        };

        Ok(Self {
            codebook: cb,
            entries,
            row_ids: row_ids.iter().map(|r| r.as_u64()).collect(),
            metric,
            dim: dim as u32,
            raw_f16,
        })
    }

    pub fn node_count(&self) -> u64 {
        self.row_ids.len() as u64
    }

    /// Search using binary codes. Returns top-k (RowId, distance) pairs sorted ascending.
    ///
    /// If `rerank_factor` > 1 and raw F16 vectors are available, the top
    /// `rerank_factor × top_k` candidates are reranked with exact F16 distances.
    ///
    /// Inner scan is intentionally sequential — callers (e.g. `SearchSession`) run
    /// shard searches in parallel via rayon. Nesting par_iter here would spawn
    /// O(shards × N) micro-tasks and cause scheduler overhead that destroys QPS.
    pub fn search(
        &self,
        query: &[f32],
        top_k: usize,
        rerank_factor: Option<usize>,
    ) -> Vec<(RowId, f32)> {
        if self.entries.is_empty() {
            return vec![];
        }
        debug_assert!(
            self.codebook.is_ready(),
            "RaBitQCodebook not initialized — call rebuild_proj() after deserialization"
        );

        let (q_proj, q_scale) = self.codebook.prepare_query(query);
        let b_q = bits_from_signs(&q_proj);
        let n = self.entries.len();
        let q_norm = query.iter().map(|x| x * x).sum::<f32>().sqrt();

        // Coarse scan: sequential — outer shard parallelism handles concurrency.
        let mut scored: Vec<(usize, f32)> = self
            .entries
            .iter()
            .enumerate()
            .map(|(i, entry)| {
                let ip = self.codebook.estimate_ip_binary(&b_q, q_scale, entry);
                let dist = match self.metric {
                    VectorMetric::Cosine | VectorMetric::NormalizedCosine => 1.0 - ip,
                    VectorMetric::DotProduct => -ip * q_norm * entry.norm,
                    VectorMetric::Euclidean => {
                        // ||q - x||² ≈ ||q||² + ||x||² - 2·ip·||q||·||x||
                        let norm_x = entry.norm;
                        (q_norm * q_norm + norm_x * norm_x - 2.0 * ip * q_norm * norm_x)
                            .max(0.0)
                            .sqrt()
                    }
                };
                (i, dist)
            })
            .collect();

        let candidates = rerank_factor.unwrap_or(1).max(1) * top_k;
        let cmp = |a: &(usize, f32), b: &(usize, f32)| {
            a.1.partial_cmp(&b.1).unwrap_or(std::cmp::Ordering::Equal)
        };
        // O(N) partial select to bring top `candidates` to front, then sort only those.
        if candidates < n {
            scored.select_nth_unstable_by(candidates - 1, cmp);
            scored.truncate(candidates);
        }
        scored.sort_unstable_by(cmp);

        // Rerank top candidates with exact F16 distances if raw vectors available
        let rerank_slice = &scored[..candidates.min(scored.len())];

        if let Some(ref raw) = self.raw_f16 {
            let dim = self.dim as usize;
            let mut reranked: Vec<(usize, f32)> = rerank_slice
                .iter()
                .map(|&(i, _)| {
                    let db_f16 = &raw[i * dim..(i + 1) * dim];
                    let db_f32: Vec<f32> = db_f16.iter().map(|x| x.to_f32()).collect();
                    let d = exact_distance(self.metric, query, &db_f32);
                    (i, d)
                })
                .collect();
            reranked.sort_unstable_by(cmp);
            reranked
                .into_iter()
                .take(top_k)
                .map(|(i, d)| (RowId::new(self.row_ids[i]), d))
                .collect()
        } else {
            rerank_slice
                .iter()
                .take(top_k)
                .map(|&(i, d)| (RowId::new(self.row_ids[i]), d))
                .collect()
        }
    }
}

// ── Serializer ────────────────────────────────────────────────────────────────

pub struct RaBitQSerializer;

impl RaBitQSerializer {
    pub fn to_bytes(index: &RaBitQIndex) -> AilakeResult<Vec<u8>> {
        bincode::serialize(index).map_err(|e| AilakeError::Bincode(e.to_string()))
    }

    pub fn from_bytes(bytes: &[u8]) -> AilakeResult<RaBitQIndex> {
        let mut idx: RaBitQIndex =
            bincode::deserialize(bytes).map_err(|e| AilakeError::Bincode(e.to_string()))?;
        idx.codebook.rebuild_proj();
        Ok(idx)
    }
}

// ── Tests ─────────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use ailake_vec::cosine_distance;
    use rand::{rngs::StdRng, Rng, SeedableRng};

    fn unit_vecs(n: usize, dim: usize, seed: u64) -> Vec<Vec<f32>> {
        let mut rng = StdRng::seed_from_u64(seed);
        (0..n)
            .map(|_| {
                let v: Vec<f32> = (0..dim).map(|_| rng.gen::<f32>() * 2.0 - 1.0).collect();
                let norm = v.iter().map(|x| x * x).sum::<f32>().sqrt();
                v.iter().map(|x| x / norm).collect()
            })
            .collect()
    }

    #[test]
    fn top1_is_self() {
        let dim = 32;
        let vecs = unit_vecs(50, dim, 1);
        let row_ids: Vec<RowId> = (0..50u64).map(RowId::new).collect();
        let idx = RaBitQIndex::build(
            &row_ids,
            &vecs,
            VectorMetric::Cosine,
            RaBitQConfig::default(),
            false,
        )
        .unwrap();

        let query = vecs[0].clone();
        let results = idx.search(&query, 1, None);
        assert_eq!(results.len(), 1);
        assert_eq!(results[0].0, RowId::new(0));
    }

    #[test]
    fn rerank_improves_recall() {
        let dim = 64;
        let n = 200;
        let vecs = unit_vecs(n, dim, 42);
        let row_ids: Vec<RowId> = (0..n as u64).map(RowId::new).collect();
        let idx = RaBitQIndex::build(
            &row_ids,
            &vecs,
            VectorMetric::Cosine,
            RaBitQConfig::default(),
            true,
        )
        .unwrap();

        let query = vecs[5].clone();

        // Brute-force ground truth
        let mut gt: Vec<(f32, u64)> = vecs
            .iter()
            .enumerate()
            .map(|(i, v)| (cosine_distance(&query, v), i as u64))
            .collect();
        gt.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());
        let gt_top5: std::collections::HashSet<u64> =
            gt.iter().take(5).map(|(_, id)| *id).collect();

        let results_reranked = idx.search(&query, 5, Some(4));
        let found: std::collections::HashSet<u64> =
            results_reranked.iter().map(|(id, _)| id.as_u64()).collect();
        let recall = found.intersection(&gt_top5).count() as f64 / 5.0;
        assert!(recall >= 0.6, "recall@5 with reranking = {recall:.2}");
    }

    #[test]
    fn serialization_roundtrip() {
        let dim = 32;
        let vecs = unit_vecs(20, dim, 7);
        let row_ids: Vec<RowId> = (0..20u64).map(RowId::new).collect();
        let idx = RaBitQIndex::build(
            &row_ids,
            &vecs,
            VectorMetric::Cosine,
            RaBitQConfig::default(),
            false,
        )
        .unwrap();

        let bytes = RaBitQSerializer::to_bytes(&idx).unwrap();
        let idx2 = RaBitQSerializer::from_bytes(&bytes).unwrap();
        let q = vecs[0].clone();
        let r1 = idx.entries[0].code.clone();
        let _r2 = idx2.search(&q, 1, None);
        assert_eq!(r1, idx2.entries[0].code);
    }
}
